domain/model/state/state.go

```diff
@@ -793,71 +788,178 @@ func (s *State) ListAllModels(ctx context.Context) ([]coremodel.Model, error) {
 		return nil, errors.Trace(err)
 	}
 
-	modelStmt := `
-SELECT uuid,
-       name,
-       cloud_name,
-       cloud_region_name,
-	   model_type,
-	   owner_uuid,
-	   owner_name,
-	   cloud_credential_name,
-	   cloud_credential_owner_name,
-	   cloud_credential_cloud_name,
-	   life
+	query := `
+SELECT &dbModel.*
 FROM v_model
 `
 
-	rval := []coremodel.Model{}
-	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		rows, err := tx.QueryContext(ctx, modelStmt)
-		if err != nil {
-			return err
+	var resultsForModel []dbModel
+
+	stmt, err := sqlair.Prepare(query, dbModel{})
+	if err != nil {
+		return nil, fmt.Errorf("preparing get model statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt).GetAll(&resultsForModel)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return fmt.Errorf("getting models from database: %w", err)
 		}
-		defer rows.Close()
-		defer func() { _ = rows.Close() }()
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting all models: %w", err)
+	}
 
-		var (
-			// cloudRegion could be null
-			cloudRegion sql.NullString
-			modelType   string
-			userUUID    string
-			credKey     credential.Key
-			model       coremodel.Model
-		)
+	return transformModelDBResult(resultsForModel...)
+}
 
-		for rows.Next() {
-			err := rows.Scan(
-				&model.UUID,
-				&model.Name,
-				&model.Cloud,
-				&cloudRegion,
-				&modelType,
-				&userUUID,
-				&model.OwnerName,
-				&credKey.Name,
-				&credKey.Owner,
-				&credKey.Cloud,
-				&model.Life,
-			)
+// HostedModels retrieves all hosted models on the controller, along with their
+// cloud and credential information. This function will filter on models that
+// have a "life" value in the includeLifes slice. Any models with an ID in the
+// excludeIDs slice will be excluded. If no matching models exist, an empty
+// slice is returned.
+func (s *State) HostedModels(
+	ctx context.Context,
+	includeLifes []corelife.Value,
+	excludeIDs []coremodel.UUID,
+) ([]coremodel.HostedModel, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT &dbModel.*
+FROM v_model
+WHERE life IN ($lifeList[:])
+AND uuid NOT IN ($modelIDList[:])
+`
+
+	inputLifes := lifeList(includeLifes)
+	inputModelIDs := modelIDList(excludeIDs)
+
+	stmt, err := sqlair.Prepare(query, dbModel{}, inputLifes, inputModelIDs)
```

Comment:
Do `s.Prepare` here. The `StateBase` has a method `Prepare` that caches the `stmt` objects, sqlair.Prepare will reprepare them each time. (We're currently working on some SQLair stuff to make this redundent but for now this is the best way to cache them)